### DISCUSSION TITLE
Web Update Q&A

### DISCUSSION THREAD

**Discussion by bernardkwok** [autodesk-forks/MaterialX#1312] [Q&A] [answered]
Feel free to ping about what's up with the current web work.
This discussion will persistent until the initial integration of ESSL and Javascript bindings is completed.

**Comment by bernardkwok** [comment, marked as answer]
For this query from @testpassword (Artemy)

> Dear Colleagues,
> 
> I am trying to start `JsMaterialXView` with my materials. I built the project with the command `npm run build` - the resulting files work fine with materials from the current repository, but when trying to open your mine material which has dependencies (files from folders `alglib`,` stdlib`, `bxdf`),` JsMaterialXGenShader.js` crashes with an error:
> 
> ```
> Error: Cycle detected!
> http://localhost:8080/bxdf/disney_brdf_2012.mtlx
> -> http://localhost:8080/bxdf/disney_brdf_2015.mtlx
> -> http://localhost:8080/stdlib/stdlib_ng.mtlx
> -> http://localhost:8080/stdlib/stdlib_defs.mtlx
> -> http://localhost:8080/bxdf/substanceMetallicRoughness.mtlx
> -> http://localhost:8080/bxdf/standard_surface.mtlx
> -> http://localhost:8080/bxdf/ubershader_defs.mtlx
> -> http://localhost:8080/bxdf/usd_preview_surface.mtlx
> -> http://localhost:8080/alglib/alglib_defs.mtlx
> -> http://localhost:8080/stdlib/stdlib_ng.mtlx
>     promise http://localhost:8080/JsMaterialXGenShader.js:9
>     promise callback*loadFiles http://localhost:8080/JsMaterialXGenShader.js:9
>     promise http://localhost:8080/JsMaterialXGenShader.js:9
>     promise callback*loadFiles http://localhost:8080/JsMaterialXGenShader.js:9
>     _readFromXmlString http://localhost:8080/JsMaterialXGenShader.js:9
>     readFromXmlString http://localhost:8080/JsMaterialXGenShader.js:9
>     init webpack://jsmaterialxview/./src/index.js?:156
>     promise callback*init webpack://jsmaterialxview/./src/index.js?:145
>     <anonymous> webpack://jsmaterialxview/./src/index.js?:38
>     js http://localhost:8080/main.js:149
>     __webpack_require__ http://localhost:8080/main.js:173
>     <anonymous> http://localhost:8080/main.js:213
>     <anonymous> http://localhost:8080/main.js:215
> index.js:241:17
>     init webpack://jsmaterialxview/./src/index.js?:241
>     (asynchronous: promise callback)
>     init webpack://jsmaterialxview/./src/index.js?:240
>     <anonymous> webpack://jsmaterialxview/./src/index.js?:38
>     js http://localhost:8080/main.js:149
>     __webpack_require__ http://localhost:8080/main.js:173
>     <anonymous> http://localhost:8080/main.js:213
>     <anonymous> http://localhost:8080/main.js:215
> ```
> 
> The main material (file `BrickWall_Red_Weathered.mtlx`) I put in the folder` resources/Materials/Examples/StandartSurface`, and dependencies (`alglib`,` bxdf`, `stdlib`,` textures`) to the build root after building `source/JsMaterialX/JsMaterialXView/dist`.
> 
> You can find the archive with the material and the full log of the browser with the error in [Google Drive](https://drive.google.com/drive/folders/1K5Lh_qMEeiB19Qy4PbPG0HAzeNeTjAw8?usp=sharing).
> 
> When viewing materials in the OS-native version of viewer, there is no problem throws, everything works correctly. I would like to know, am I doing something wrong, or are materials with dependencies not supported at the moment?
> 
> Thanks for the help

The current viewer setup has only been tested with the existing librarys and hence the Examples folder materials.

It seems what you've done here is add in the `alglib` (looks like the Substance Designer library ?).
From the logs it looks like it's trying to reload in the same library files more than once which could be the problem but hard to say without having those files to try out.

This is a good litmus test though that we can look at expanding this a bit so that any dropped in library which is packaged for the
demo view works with a sample material. Will add this as a item to examine for the Javacript work.

Hope this helps,
Bernard

**Comment by testpassword** [comment]
Thanks for rapid reply. Yes, it's Substance 3D Designer. For now, all materials created in it do not work. Then I'll try to fix it, if it works, I'll let you know.

Good luck in development,
Artemy

**Comment by bernardkwok** [comment]
Hi Artemy,

Thanks for looking at this. Let us know who it works out.

We've wanted to maintain that MTLX exported from Substance Designer works with MaterialXView, and
going forward we'd want to maintain this for any sample web based viewer. One thing that was discussed with Adobe before is that there can be incompatibilities if libraries such as stdlib, pbrlib as also included in the MaterialX export (you get 2 copies), and as  I recall there is an option on the plug-in to not do this extra export. 

BTW, 
- https://github.com/autodesk-forks/MaterialX/blob/adsk_contrib/dev/resources/Materials/Examples/StandardSurface/standard_surface_brick_procedural.mtlx
- https://github.com/materialx/MaterialX/blob/main/resources/Materials/Examples/StandardSurface/standard_surface_brick_procedural.mtlx
(same file) is something we co-worked on and is exported from Designer.
